README.md
# Prometheus metrics aggregator [![Latest Release](https://img.shields.io/github/release/peterbourgon/prometheus-aggregator.svg?style=flat-square)](https://github.com/peterbourgon/prometheus-aggregator/releases/latest) [![Build Status](https://img.shields.io/endpoint.svg?url=https%3A%2F%2Factions-badge.atrox.dev%2Fpeterbourgon%2Fprometheus-aggregator%2Fbadge%3Fref%3Dmain&style=flat)](https://actions-badge.atrox.dev/peterbourgon/prometheus-aggregator/goto?ref=main)

Receive and aggregate metrics for consumption by a Prometheus server.

**DON'T USE THIS TOOL**. If at all possible, you should expose a /metrics
endpoint in your service and have Prometheus scrape it directly. If you're
running a cron job, prefer [the pushgateway][pushgateway]. This tool only exists
to help with edge case scenarios, for example Perl web services that use a
[Unicorn][unicorn]-style forked process model to handle concurrency, and are
difficult or impossible to get Prometheus to scrape.

Related work:

- [prometheus/pushgateway][pushgateway] doesn't do aggregation but works OK for things like batch or cron jobs
- [prometheus/statsd_exporter][statsd] accepts StatsD writes, but requires a big YAML config for mappings
- [weaveworks/prom-aggregation-gateway][pag] accepts HTTP POSTs from its corresponding [JavaScript client][jsc]
- [discourse/prometheus_exporter][dpe] is a Ruby solution with client gem, but no histograms

[pushgateway]: https://github.com/prometheus/pushgateway
[unicorn]: https://bogomips.org/unicorn/
[statsd]: https://github.com/prometheus/statsd_exporter/
[pag]: https://github.com/weaveworks/prom-aggregation-gateway/
[jsc]: https://github.com/weaveworks/promjs/
[dpe]: https://github.com/discourse/prometheus_exporter/

## Getting

[Download the latest release][release] if you're lazy, or build it yourself from
latest master if you have the Go toolchain installed and have YOLO tattooed on
your knuckles or whatever.

[release]: https://github.com/peterbourgon/prometheus-aggregator/releases/latest

```
go get github.com/peterbourgon/prometheus-aggregator
$GOPATH/bin/prometheus-aggregator -h
```

## Running

I mean you just kinda run it.

```
USAGE
  prometheus-aggregator [flags]

FLAGS
  -debug false                              log debug information
  -declfile ...                             file containing JSON metric declarations
  -declpath ...                             sibling path to /metrics serving declfile contents
  -example false                            print example declfile to stdout and return
  -prometheus tcp://127.0.0.1:8192/metrics  address for Prometheus scrapes
  -socket tcp://127.0.0.1:8191              address for direct socket metric writes
  -strict false                             disconnect clients when they send bad data

VERSION
  0.0.15
```

## How it works

The prometheus-aggregator expects clients to connect and emit newline-delimited
JSON objects for each metric observation. Each object needs to be fully
specified with name, type, help, and value. Here's an example of three counter
increments.

```
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos.", "value": 1}
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos.", "value": 1}
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos.", "value": 1}
```

Obviously this is wildly inefficient, so, as an optimization, once a metric has
been "declared" with name, type, and help, subsequent emissions may refer to it
simply by name.

```
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos.", "value": 1}
{"name": "myapp_foo_total", "value": 1}
{"name": "myapp_foo_total", "value": 1}
```

You can delare a metric without making an observation by omitting the value.

```
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos."}
{"name": "myapp_foo_total", "value": 1}  # value is now 1
{"name": "myapp_foo_total", "value": 2}  # value is now 3
```

You can declare metrics at runtime, like this, or you can predeclare metrics in
a file containing a JSON array of multiple JSON objects, and pass it to the
program at startup via the `-declfile` flag. Or mix and match both! Life is
full of possibility.

New! Exciting! Great Value! An optional `-declpath` flag allows you to serve
your initial metric declarations on a sibling path to your Prometheus metrics
telemetry. This can be useful if you want to programmatically verify the state
of a prometheus-aggregator instance.

## Prometheus exposition format

If serializing JSON is a bottleneck, you can optionally emit observations (but
not declarations) in the [Prometheus exposition format][pef]. Note that the
parser (such as it is) is pretty strict, so don't get crazy with whitespace or
whatever.

[pef]: https://prometheus.io/docs/instrumenting/exposition_formats/

```
{"name": "myapp_foo_total", "type": "counter", "help": "Total number of foos."}
myapp_foo_total{} 1
myapp_foo_total{} 2
```

## Labels

Labels are supported in both formats as you might expect.

```
{"name": "myapp_foo_total", "labels": {"success": "false", "code": "401"}, "value": 1}
myapp_foo_total{success="true",code="200"} 1
```

## Supported types

Counters are obviously supported. Gauges are also supported and work just like
counters, but default to setting themselves to the most recent value.

```
{"name": "myapp_worker_pool", "type": "gauge", "help": "Size of worker pool."}
myapp_worker_pool{} 1  # value is now 1
myapp_worker_pool{} 2  # value is now 2
```

Histograms are supported too. Provide buckets with the declaration.

```
{"name": "myapp_req_dur_seconds", "type": "histogram",
  "help": "Duration of request in seconds.",
    "buckets": [0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10]}
myapp_req_dur_seconds{} 0.0123
myapp_req_dur_seconds{} 0.99
```

**Summaries are not supported**. This is fine, you can't do meaningful
aggregation over summaries at query time anyway. You'll need to define some
buckets and I know that sounds hard, and it _is_ hard, life is hard, I'm sorry
for that.

## Bad data

By default, if a client sends bad data, the only thing that happens is the
prometheus-aggregator will log an error, the client won't know about it. This is
a good idea for production but maybe for dev you want to pass the `-strict`
flag, which means if a client sends bad data it gets disconnected!! Harsh!!

## UDP

You can specify a socket write address as e.g. `udp://127.0.0.1:8191` and then
you can emit UDP observations! The same rules apply, one metric per datagram.
The `-strict` flag has no meaning in this mode as UDP is connectionless.


CHANGELOG.md
## v0.0.15

- Fix crashing bug with UDP -socket addresses.

## v0.0.14

- Add `-declpath` flag.

## v0.0.13

- First tagged version that is modules-aware.
- Stop vendoring dependencies.
- Switch to GitHub Actions for CI.
- Minor fixes to satisfy new static analysis tools.

## v0.0.12

- A typo in an error message when parsing Prometheus exposition format lines has been corrected.

## v0.0.11

- Bug fix: don't render timeseries that haven't collected any observations.

## v0.0.10

- Make histograms abide the convention of using `_bucket` suffixes on the metric name.

## v0.0.9

- Fix bug when parsing UNIX domain sockets.

## v0.0.8

- Add UDP support for incoming socket writes.

## v0.0.7

- Fixes a crashing bug if you don't specify an explicit -prometheus path.

## v0.0.6

- Renamed -direct to -socket.

## v0.0.5

- Removed discourse/prometheus_exporter compatibility.
- It was a bad idea.

## v0.0.4

- Added discourse/prometheus_exporter compatibility.
- This might be a good idea!

## v0.0.3

- Fixed tests.

## v0.0.2

- Emit HELP before TYPE in Prometheus /metrics endpoint.

## v0.0.1

- Initial release.



full_test.go
package main

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestThreeTypes(t *testing.T) {
	u, _ := newUniverse()
	loadObservations(t, u, makeObservations(t, []string{
		`{"name":"foo_total","type":"counter","help":"Total number of foos."}`,
		`{"name":"foo_total","labels":{"code":"200"},"value": 1}`,
		`{"name":"foo_total","labels":{"code":"404"},"value": 2}`,
		`foo_total{code="200"} 4`,
		`foo_total{code="404"} 8`,

		`{"name":"bar_seconds","type":"histogram","help":"Bar duration in seconds.","buckets":[0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10]}`,
		`{"name":"bar_seconds","value":0.123}`,
		`{"name":"bar_seconds","value":0.234}`,
		`{"name":"bar_seconds","value":0.501}`,
		`{"name":"bar_seconds","value":8.000}`,

		`{"name":"baz_size","type":"gauge","help":"Current size of baz widget."}`,
		`{"name":"baz_size","value": 1}`,
		`{"name":"baz_size","value": 2}`,
		`baz_size{} 4`,
	}))
	if want, have := normalizeResponse(`
		# HELP bar_seconds Bar duration in seconds.
		# TYPE bar_seconds histogram
		bar_seconds_bucket{le="0.01"} 0
		bar_seconds_bucket{le="0.05"} 0
		bar_seconds_bucket{le="0.1"} 0
		bar_seconds_bucket{le="0.5"} 2
		bar_seconds_bucket{le="1"} 3
		bar_seconds_bucket{le="2"} 3
		bar_seconds_bucket{le="5"} 3
		bar_seconds_bucket{le="10"} 4
		bar_seconds_bucket{le="+Inf"} 4
		bar_seconds_sum{} 8.858000
		bar_seconds_count{} 4
		
		# HELP baz_size Current size of baz widget.
		# TYPE baz_size gauge
		baz_size{} 4.000000
		
		# HELP foo_total Total number of foos.
		# TYPE foo_total counter
		foo_total{code="200"} 5.000000
		foo_total{code="404"} 10.000000
	`), normalizeResponse(scrape(t, u)); want != have {
		t.Fatalf("\n---WANT---\n%s\n\n---HAVE---\n%s\n", want, have)
	}
}

func TestInitialDeclarations(t *testing.T) {
	u, _ := newUniverse(makeObservations(t, []string{
		`{"name":"foo_total","type":"counter","help":"Total number of foos."}`,
		`{"name":"bar_seconds","type":"histogram","help":"Bar duration in seconds.","buckets":[0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10]}`,
		`{"name":"baz_size","type":"gauge","help":"Current size of baz widget."}`,
		`{"name":"qux_count","type":"counter","help":"Count of qux events."}`,
	})...)
	loadObservations(t, u, makeObservations(t, []string{
		`foo_total{label="value"} 1`,
		`bar_seconds{} 0.234`,
		`baz_size{} 5`,
	}))
	if want, have := normalizeResponse(`
		# HELP bar_seconds Bar duration in seconds.
		# TYPE bar_seconds histogram
		bar_seconds_bucket{le="0.01"} 0
		bar_seconds_bucket{le="0.05"} 0
		bar_seconds_bucket{le="0.1"} 0
		bar_seconds_bucket{le="0.5"} 1
		bar_seconds_bucket{le="1"} 1
		bar_seconds_bucket{le="2"} 1
		bar_seconds_bucket{le="5"} 1
		bar_seconds_bucket{le="10"} 1
		bar_seconds_bucket{le="+Inf"} 1
		bar_seconds_sum{} 0.234000
		bar_seconds_count{} 1
		
		# HELP baz_size Current size of baz widget.
		# TYPE baz_size gauge
		baz_size{} 5.000000
		
		# HELP foo_total Total number of foos.
		# TYPE foo_total counter
		foo_total{label="value"} 1.000000
	`), normalizeResponse(scrape(t, u)); want != have {
		t.Fatalf("\n---WANT---\n%s\n\n---HAVE---\n%s\n", want, have)
	}
}

func makeObservations(t *testing.T, lines []string) []observation {
	t.Helper()
	observations := make([]observation, len(lines))
	for i, s := range lines {
		o, err := parseLine([]byte(s))
		if err != nil {
			t.Fatal(err)
		}
		observations[i] = o
	}
	return observations
}

func loadObservations(t *testing.T, obs observer, observations []observation) {
	t.Helper()
	for _, o := range observations {
		if err := obs.observe(o); err != nil {
			t.Fatalf("%+v: %v", o, err)
		}
	}
}

func scrape(t *testing.T, h http.Handler) string {
	rec := httptest.NewRecorder()
	req, _ := http.NewRequest("GET", "/", nil)
	h.ServeHTTP(rec, req)
	return rec.Body.String()
}

func normalizeResponse(s string) string {
	lines := strings.Split(s, "\n")
	for i := range lines {
		lines[i] = strings.TrimSpace(lines[i])
	}
	return strings.TrimSpace(strings.Join(lines, "\n"))
}


parse_test.go
package main

import (
	"testing"

	"github.com/google/go-cmp/cmp"
)

func TestParsePrometheus(t *testing.T) {
	fp := func(f float64) (p *float64) {
		p = new(float64)
		*p = f
		return p
	}

	for name, testcase := range map[string]struct {
		input string
		obs   observation
		err   bool
	}{
		"only spaces": {
			input: `    `,
			err:   true,
		},
		"empty": {
			input: ``,
			err:   true,
		},
		"no braces": {
			input: `foo 1`,
			err:   true,
		},
		"leading space": {
			input: ` foo{} 1`,
			obs:   observation{Name: "foo", Value: fp(1.00), Labels: map[string]string{}},
		},
		"trailing space": {
			input: `foo{} 1 `,
			obs:   observation{Name: "foo", Value: fp(1.00), Labels: map[string]string{}},
		},
		"ascii value": {
			input: `foo{} A`,
			err:   true,
		},
		"most basic": {
			input: `foo{} 1`,
			obs:   observation{Name: "foo", Value: fp(1.00), Labels: map[string]string{}},
		},
		"with label": {
			input: `foo{code="200"} 2.34`,
			obs:   observation{Name: "foo", Value: fp(2.34), Labels: map[string]string{"code": "200"}},
		},
		"missing quotes": {
			input: `foo{code=200} 2.34`,
			err:   true,
		},
		"missing closing brace": {
			input: `foo{code="200" 7`,
			err:   true,
		},
		"double closing brace": {
			input: `foo{code="200"}} 7`,
			err:   true,
		},
		"two labels": {
			input: `foo{code="200",err="false"} 7`,
			obs:   observation{Name: "foo", Value: fp(7.00), Labels: map[string]string{"code": "200", "err": "false"}},
		},
		"space between labels": {
			input: `foo{code="200", err="false"} 7`,
			err:   true,
		},
		"space instead of comma": {
			input: `foo{code="200" err="false"} 7`,
			err:   true,
		},
	} {
		t.Run(name, func(t *testing.T) {
			var obs observation
			err := prometheusUnmarshal([]byte(testcase.input), &obs)
			if want, have := testcase.err, err != nil; want != have {
				t.Fatalf("err: want %v, have %v (%v)", want, have, err)
			}
			if want, have := testcase.obs, obs; !cmp.Equal(want, have) {
				t.Fatal(cmp.Diff(want, have))
			}
		})
	}
}

func TestParseURI(t *testing.T) {
	for name, testcase := range map[string]struct {
		input   string
		network string
		address string
		path    string
		err     bool
	}{
		"empty string": {
			input: "",
			err:   true,
		},
		"bad scheme": {
			input: ":/localhost:1234/x",
			err:   true,
		},
		"TCP with path": {
			input:   "tcp://localhost:1234/x",
			network: "tcp",
			address: "localhost:1234",
			path:    "/x",
		},
		"TCP4 without path": {
			input:   "tcp4://localhost:1234",
			network: "tcp4",
			address: "localhost:1234",
			path:    "",
		},
		"UNIX socket": {
			input:   "unix:///var/tmp/my.sock",
			network: "unix",
			address: "/var/tmp/my.sock",
			path:    "",
		},
		"UNIX socket with host": {
			input:   "unix://ignored-host/var/tmp/my.sock",
			network: "unix",
			address: "/var/tmp/my.sock",
			path:    "",
		},
	} {
		t.Run(name, func(t *testing.T) {
			network, address, path, err := parseURI(testcase.input)
			if testcase.err && err == nil {
				t.Fatal("wanted error, but got none")
			}
			if want, have := testcase.network, network; want != have {
				t.Errorf("network: want %q, have %q", want, have)
			}
			if want, have := testcase.address, address; want != have {
				t.Errorf("address: want %q, have %q", want, have)
			}
			if want, have := testcase.path, path; want != have {
				t.Errorf("path: want %q, have %q", want, have)
			}
		})
	}
}


write_test.go
package main

import (
	"fmt"
	"io"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/go-kit/kit/log"
)

func TestHandleConn(t *testing.T) {
	// Set up our little universe.
	var (
		dst, _ = newUniverse()
		src, w = io.Pipe()
		strict = true
		logger = log.NewNopLogger()
	)

	// Take writes from the output of the pipe into the universe.
	done := make(chan struct{})
	go func() {
		defer close(done)
		handleConn(src, dst, strict, logger)
	}()

	// Make writes to the input of the pipe.
	fmt.Fprintln(w, `{"name":"foo","type":"counter","help":"Total foos.","labels":{"code":"412"},"value":1}`)
	fmt.Fprintln(w, `{"name":"foo","labels":{"code":"412"},"value":2}`)
	fmt.Fprintln(w, `foo{code="412"} 4`)

	// Close the pipe, and wait for the handleDirectWrites goroutine to exit.
	w.Close()
	<-done

	// Dump the Prometheus /metrics output, and verify it.
	rec := httptest.NewRecorder()
	dst.ServeHTTP(rec, &http.Request{})
	if want, have := normalizeResponse(`
		# HELP foo Total foos.
		# TYPE foo counter
		foo{code="412"} 7.000000
	`), normalizeResponse(rec.Body.String()); want != have {
		t.Fatalf("\n---WANT---\n%s\n\n---HAVE---\n%s\n", want, have)
	}
}


handle_line.go
package main

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"strconv"

	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/log/level"
	"github.com/pkg/errors"
)

type observer interface{ observe(observation) error }

func forwardPacketConn(conn net.PacketConn, o observer, logger log.Logger) error {
	buf := make([]byte, bufio.MaxScanTokenSize)
	for {
		n, _, err := conn.ReadFrom(buf)
		if err != nil {
			return err
		}
		name, err := handleLine(buf[:n], o)
		if err != nil {
			level.Error(logger).Log("line", "rejected", "err", err)
			continue
		}
		level.Debug(logger).Log("line", "accepted", "name", name)
	}
}

func forwardListener(ln net.Listener, o observer, strict bool, logger log.Logger) error {
	for {
		conn, err := ln.Accept()
		if err != nil {
			return err
		}
		go handleConn(conn, o, strict, log.With(logger, "remote_addr", conn.RemoteAddr()))
	}
}

func handleConn(rc io.ReadCloser, o observer, strict bool, logger log.Logger) {
	defer rc.Close()
	s := bufio.NewScanner(rc)
	for s.Scan() {
		name, err := handleLine(s.Bytes(), o)
		if err != nil {
			level.Error(logger).Log("line", "rejected", "err", err)
			if strict {
				return
			}
			continue
		}
		level.Debug(logger).Log("line", "accepted", "name", name)
	}
}

func handleLine(line []byte, o observer) (string, error) {
	obs, err := parseLine(line)
	if err != nil {
		return "", errors.Wrap(err, "parse error")
	}
	if err := o.observe(obs); err != nil {
		return obs.Name, errors.Wrap(err, "observation error")
	}
	return obs.Name, nil
}

func parseLine(p []byte) (o observation, err error) {
	if len(p) <= 0 {
		err = errors.New("invalid (empty) line")
	} else if p[0] == '{' {
		err = json.Unmarshal(p, &o)
	} else {
		err = prometheusUnmarshal(p, &o)
	}
	return o, err
}

func prometheusUnmarshal(p []byte, o *observation) error {
	p = bytes.TrimSpace(p)
	x := bytes.LastIndexByte(p, ' ')
	if x < 1 {
		return fmt.Errorf("bad format: couldn't find space")
	}

	id, val := bytes.TrimSpace(p[:x]), bytes.TrimSpace(p[x+1:])

	value, err := strconv.ParseFloat(string(val), 64)
	if err != nil {
		return errors.Wrapf(err, "bad value (%s)", string(val))
	}

	y := bytes.IndexByte(id, '{')
	if y < 0 {
		return fmt.Errorf("bad format: couldn't find opening brace")
	}
	if id[len(id)-1] != '}' {
		return fmt.Errorf("bad format: couldn't find terminating brace")
	}

	name, labels := id[:y], id[y+1:len(id)-1]
	if bytes.ContainsRune(labels, ' ') {
		return fmt.Errorf("bad format: labels section may not contain spaces")
	}

	labelmap := map[string]string{}
	for _, pair := range bytes.Split(labels, []byte(",")) {
		z := bytes.IndexByte(pair, '=')
		if z < 0 {
			continue
		}
		k, v := pair[:z], pair[z+1:]
		if v[0] != '"' || v[len(v)-1] != '"' {
			return fmt.Errorf("bad format: label value must be wrapped in quotes")
		}
		v = v[1 : len(v)-1]
		labelmap[string(k)] = string(v)
	}

	o.Name = string(name)
	o.Labels = labelmap
	o.Value = new(float64)
	(*o.Value) = value

	return nil
}


main.go
package main

import (
	"context"
	"encoding/json"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"strings"
	"syscall"
	"text/tabwriter"
	"time"

	"github.com/go-kit/kit/log"
	"github.com/go-kit/kit/log/level"
	"github.com/oklog/run"
)

var version = "HEAD (dev/unreleased)"

func main() {
	fs := flag.NewFlagSet("prometheus-aggregator", flag.ExitOnError)
	var (
		sockAddr = fs.String("socket", "tcp://127.0.0.1:8191", "address for direct socket metric writes")
		promAddr = fs.String("prometheus", "tcp://127.0.0.1:8192/metrics", "address for Prometheus scrapes")
		declfile = fs.String("declfile", "", "file containing JSON metric declarations")
		declpath = fs.String("declpath", "", "sibling path to /metrics serving declfile contents")
		example  = fs.Bool("example", false, "print example declfile to stdout and return")
		debug    = fs.Bool("debug", false, "log debug information")
		strict   = fs.Bool("strict", false, "disconnect clients when they send bad data")
	)
	fs.Usage = usageFor(fs, "prometheus-aggregator [flags]")
	fs.Parse(os.Args[1:])

	if *example {
		buf, _ := json.MarshalIndent(exampleDecls, "", "    ")
		fmt.Fprintf(os.Stdout, "%s\n", buf)
		os.Exit(0)
	}

	var logger log.Logger
	{
		logger = log.NewLogfmtLogger(os.Stdout)
		loglevel := level.AllowInfo()
		if *debug {
			loglevel = level.AllowDebug()
		}
		logger = level.NewFilter(logger, loglevel)
	}

	var initial []observation
	{
		if *declfile != "" {
			buf, err := ioutil.ReadFile(*declfile)
			if err != nil {
				level.Error(logger).Log("err", err)
				os.Exit(1)
			}
			if err := json.Unmarshal(buf, &initial); err != nil {
				level.Error(logger).Log("err", err)
				os.Exit(1)
			}
		}
	}

	var u *universe
	{
		var err error
		u, err = newUniverse(initial...)
		if err != nil {
			level.Error(logger).Log("err", err)
			os.Exit(1)
		}
	}

	var socketNetwork, socketAddress string
	var forwardFunc func() error
	var forwardClose func() error
	{
		sockURL, err := url.Parse(*sockAddr)
		if err != nil {
			level.Error(logger).Log("socket", *sockAddr, "err", err)
			os.Exit(1)
		}

		socketNetwork = strings.ToLower(sockURL.Scheme)
		switch socketNetwork {
		case "udp", "udp4", "udp6", "tcp", "tcp4", "tcp6":
			socketAddress = sockURL.Host
		case "unix", "unixgram", "unipacket":
			socketAddress = sockURL.Path
		default:
			level.Error(logger).Log("socket", *sockAddr, "err", "unsupported network", "network", sockURL.Scheme)
			os.Exit(1)
		}

		switch socketNetwork {
		case "udp", "udp4", "udp6", "unixgram":
			laddr, err := net.ResolveUDPAddr(sockURL.Scheme, socketAddress)
			if err != nil {
				level.Error(logger).Log("socket", *sockAddr, "err", err)
				os.Exit(1)
			}
			conn, err := net.ListenUDP(sockURL.Scheme, laddr)
			if err != nil {
				level.Error(logger).Log("socket", *sockAddr, "err", err)
				os.Exit(1)
			}
			forwardFunc = func() error { return forwardPacketConn(conn, u, logger) }
			forwardClose = conn.Close

		case "tcp", "tcp4", "tcp6", "unix", "unixpacket":
			ln, err := net.Listen(sockURL.Scheme, socketAddress)
			if err != nil {
				level.Error(logger).Log("socket", *sockAddr, "err", err)
				os.Exit(1)
			}
			forwardFunc = func() error { return forwardListener(ln, u, *strict, logger) }
			forwardClose = ln.Close
		}
	}

	var metricsLn net.Listener
	var metricsPath string
	{
		u, err := url.Parse(*promAddr)
		if err != nil {
			level.Error(logger).Log("prometheus", *promAddr, "err", err)
			os.Exit(1)
		}
		metricsLn, err = net.Listen(u.Scheme, u.Host)
		if err != nil {
			level.Error(logger).Log("prometheus", *promAddr, "err", err)
			os.Exit(1)
		}
		metricsPath = u.Path
		if metricsPath == "" {
			metricsPath = "/"
		}
	}

	var declPath string
	var declHandler http.Handler
	{
		if *declpath != "" {
			*declpath = "/" + strings.Trim(*declpath, "/ ")
			scheme := metricsLn.Addr().Network()
			hostport := metricsLn.Addr().String()
			u, err := url.Parse(scheme + "://" + hostport + *declpath)
			if err != nil {
				level.Error(logger).Log("declpath", *declpath, "err", err)
				os.Exit(1)
			}
			declPath = u.Path
			response, err := json.MarshalIndent(initial, "", "    ")
			if err != nil {
				level.Error(logger).Log("declpath", *declpath, "during", "marshal of initial declarations", "err", err)
				os.Exit(1)
			}
			declHandler = http.HandlerFunc(func(w http.ResponseWriter, _ *http.Request) {
				w.Header().Set("content-type", "application/json; charset=utf-8")
				w.Write(response)
			})
		}
	}

	var g run.Group
	{
		g.Add(func() error {
			level.Info(logger).Log("listener", "socket_writes", "network", socketNetwork, "address", socketAddress)
			return forwardFunc()
		}, func(error) {
			forwardClose()
		})
	}
	{
		mux := http.NewServeMux()
		mux.Handle(metricsPath, u)
		if declPath != "" {
			mux.Handle(declPath, declHandler)
		}
		server := http.Server{Handler: mux}
		g.Add(func() error {
			keyvals := []interface{}{"listener", "prometheus_scrapes", "network", metricsLn.Addr().Network(), "address", metricsLn.Addr().String(), "path", metricsPath}
			if declPath != "" {
				keyvals = append(keyvals, "declarations", declPath)
			}
			level.Info(logger).Log(keyvals...)
			return server.Serve(metricsLn)
		}, func(error) {
			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
			defer cancel()
			if err := server.Shutdown(ctx); err != nil {
				level.Error(logger).Log("err", err)
			}
		})
	}
	{
		ctx, cancel := context.WithCancel(context.Background())
		g.Add(func() error {
			c := make(chan os.Signal, 1)
			signal.Notify(c, syscall.SIGINT, syscall.SIGTERM)
			select {
			case sig := <-c:
				return fmt.Errorf("received signal %s", sig)
			case <-ctx.Done():
				return ctx.Err()
			}
		}, func(error) {
			cancel()
		})
	}
	level.Info(logger).Log("exit", g.Run())
}

func usageFor(fs *flag.FlagSet, short string) func() {
	return func() {
		fmt.Fprintf(os.Stderr, "USAGE\n")
		fmt.Fprintf(os.Stderr, "  %s\n", short)
		fmt.Fprintf(os.Stderr, "\n")
		fmt.Fprintf(os.Stderr, "FLAGS\n")
		w := tabwriter.NewWriter(os.Stderr, 0, 2, 2, ' ', 0)
		fs.VisitAll(func(f *flag.Flag) {
			def := f.DefValue
			if def == "" {
				def = "..."
			}
			fmt.Fprintf(w, "\t-%s %s\t%s\n", f.Name, def, f.Usage)
		})
		w.Flush()
		fmt.Fprintf(os.Stderr, "\n")
		fmt.Fprintf(os.Stderr, "VERSION\n")
		fmt.Fprintf(os.Stderr, "  %s\n", version)
		fmt.Fprintf(os.Stderr, "\n")
	}
}

var exampleDecls = []observation{
	{
		Name: "myservice_jobs_processed_total",
		Type: "counter",
		Help: "Total number of jobs processed.",
	},
	{
		Name: "myservice_cache_size_bytes",
		Type: "gauge",
		Help: "Current size of cache in bytes.",
	},
	{
		Name:    "myservice_http_request_duration_seconds",
		Type:    "histogram",
		Help:    "HTTP request duraton in seconds.",
		Buckets: []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10},
	},
}
